lib/block_view_lockups.go

```diff
@@ -818,109 +1174,541 @@ func (bav *UtxoView) _connectCoinLockup(
+	var previousLockedBalanceEntry *LockedBalanceEntry
+	var previousLockedBalanceEntries []*LockedBalanceEntry
+	if txMeta.UnlockTimestampNanoSecs == txMeta.VestingEndTimestampNanoSecs {
+		// Unvested consolidation case:
+
+		// (1) Check for a locked balance entry with the same unlock time
+		lockedBalanceEntry, err := bav.GetLockedBalanceEntryForLockedBalanceEntryKey(
+			&LockedBalanceEntryKey{
+				HODLerPKID:                  *hodlerPKID,
+				ProfilePKID:                 *profilePKID,
+				UnlockTimestampNanoSecs:     txMeta.UnlockTimestampNanoSecs,
+				VestingEndTimestampNanoSecs: txMeta.VestingEndTimestampNanoSecs,
+			})
+		if err != nil {
+			return 0, 0, nil,
+				errors.Wrap(err, "_connectCoinLockup failed to fetch unvested lockedBalanceEntry")
+		}
+		if lockedBalanceEntry == nil || lockedBalanceEntry.isDeleted {
+			lockedBalanceEntry = &LockedBalanceEntry{
+				HODLerPKID:                  hodlerPKID,
+				ProfilePKID:                 profilePKID,
+				UnlockTimestampNanoSecs:     txMeta.UnlockTimestampNanoSecs,
+				VestingEndTimestampNanoSecs: txMeta.VestingEndTimestampNanoSecs,
+				BalanceBaseUnits:            *uint256.NewInt(),
+			}
 		}
-	}
-	previousLockedBalanceEntry := *lockedBalanceEntry
 
-	// Attempt to add the value of this lockup transaction to the locked balance entry.
-	newLockedBalanceEntryBalance, err := SafeUint256().Add(&lockedBalanceEntry.BalanceBaseUnits, lockupValue)
-	if err != nil {
-		return 0, 0, nil,
-			errors.Wrap(RuleErrorCoinLockupYieldCausesOverflowInLockedBalanceEntry,
-				"_connectCoinLockup: New Locked Balance Entry Balance")
-	}
+		// (2) Store the previous locked balance entry
+		previousLockedBalanceEntry = lockedBalanceEntry.Copy()
 
-	// Ensure in the case of DESO the resulting locked amount is less than 2**64.
-	if profilePKID.IsZeroPKID() && !newLockedBalanceEntryBalance.IsUint64() {
-		return 0, 0, nil,
-			errors.Wrap(RuleErrorCoinLockupYieldCausesOverflow, "_connectCoinLockup: "+
-				"New DESO Locked Balance Entry Balance")
-	}
+		// (3) Check for consolidation overflow
+		newLockedBalanceEntryBalance, err := SafeUint256().Add(&lockedBalanceEntry.BalanceBaseUnits, lockupValue)
+		if err != nil {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupYieldCausesOverflowInLockedBalanceEntry,
+					"_connectCoinLockup: New Locked Balance Entry Balance")
+		}
+		// Ensure in the case of DESO the resulting locked amount is less than 2**64.
+		if profilePKID.IsZeroPKID() && !newLockedBalanceEntryBalance.IsUint64() {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupYieldCausesOverflow, "_connectCoinLockup: "+
+					"New DESO Locked Balance Entry Balance")
+		}
+
+		// (4) Set the new locked balance entry in the view
+		lockedBalanceEntry.BalanceBaseUnits = *newLockedBalanceEntryBalance
+		bav._setLockedBalanceEntry(lockedBalanceEntry)
+
+		// NOTE: While we could check for "global" overflow here, we let this occur on the unlock transaction instead.
```

Comment:
Very interesting edge-case for yields. I have to think more about if we're covered here, but I think we're OK. We're definitely fine for coins that skip yield, which is good.